arwen/interface.go

```diff
@@ -13,96 +24,137 @@ const (
-	GetStorage(addr []byte, key []byte) []byte
-	SetStorage(addr []byte, key []byte, value []byte) int32
+type BlockchainContext interface {
+	NewAddress(creatorAddress []byte) ([]byte, error)
+	AccountExists(addr []byte) bool
 	GetBalance(addr []byte) []byte
-	BlockHash(nonce int64) []byte
-	GetVMInput() vmcommon.VMInput
-	GetSCAddress() []byte
-	WriteLog(addr []byte, topics [][]byte, data []byte)
-	Transfer(destination []byte, sender []byte, gasLimit uint64, value *big.Int, input []byte)
-	Finish(data []byte)
-	BlockChainHook() vmcommon.BlockchainHook
-	SignalUserError()
-	ReturnData() [][]byte
+	GetNonce(addr []byte) (uint64, error)
+	CurrentEpoch() uint32
+	GetStateRootHash() []byte
+	LastTimeStamp() uint64
+	LastNonce() uint64
+	LastRound() uint64
+	LastEpoch() uint32
+	CurrentRound() uint64
+	CurrentNonce() uint64
+	CurrentTimeStamp() uint64
+	CurrentRandomSeed() []byte
+	LastRandomSeed() []byte
+	IncreaseNonce(addr []byte)
+	GetCodeHash(addr []byte) ([]byte, error)
+	GetCode(addr []byte) ([]byte, error)
+	GetCodeSize(addr []byte) (int32, error)
+	BlockHash(number int64) []byte
+}
 
+type RuntimeContext interface {
+	StateStack
+
+	InitStateFromContractCallInput(input *vmcommon.ContractCallInput)
+	GetVMInput() *vmcommon.VMInput
+	SetVMInput(vmInput *vmcommon.VMInput)
+	GetSCAddress() []byte
+	SetSCAddress(scAddress []byte)
+	GetVMType() []byte
+	Function() string
+	Arguments() [][]byte
+	SignalUserError(message string)
+	SignalExit(exitCode int)
 	SetRuntimeBreakpointValue(value BreakpointValue)
 	GetRuntimeBreakpointValue() BreakpointValue
-
+	PushInstance()
+	PopInstance() error
+	ReadOnly() bool
 	SetReadOnly(readOnly bool)
-	CreateNewContract(input *vmcommon.ContractCreateInput) ([]byte, error)
-	ExecuteOnSameContext(input *vmcommon.ContractCallInput) error
-	ExecuteOnDestContext(input *vmcommon.ContractCallInput) error
+	CreateWasmerInstance(contract []byte) error
+	CreateWasmerInstanceWithGasLimit(contract []byte, gasLimit uint64) error
+	SetInstanceContext(instCtx *wasmer.InstanceContext)
+	GetInstanceContext() *wasmer.InstanceContext
+	GetInstanceExports() wasmer.ExportsMap
+	GetInitFunction() wasmer.ExportedFunctionCallback
+	GetFunctionToCall() (wasmer.ExportedFunctionCallback, error)
+	GetPointsUsed() uint64
+	SetPointsUsed(gasPoints uint64)
+	MemStore(offset int32, data []byte) error
+	MemLoad(offset int32, length int32) ([]byte, error)
+	CleanInstance()
+	SetInstanceContextId(id int)
 }
 
 type BigIntContext interface {
-	GasSchedule() *config.GasCost
-	UseGas(gas uint64)
-	GasLeft() uint64
+	StateStack
+
 	Put(value int64) int32
 	GetOne(id int32) *big.Int
 	GetTwo(id1, id2 int32) (*big.Int, *big.Int)
 	GetThree(id1, id2, id3 int32) (*big.Int, *big.Int, *big.Int)
-	Arguments() [][]byte
-	GetStorage(addr []byte, key []byte) []byte
-	SetStorage(addr []byte, key []byte, value []byte) int32
-	GetVMInput() vmcommon.VMInput
-	GetBalance(addr []byte) []byte
-	GetSCAddress() []byte
+}
+
+// TODO find a better name
```

Comment:
I think it's fine. But if you want something else, it can also be ResultContext, or TxResultContext